File: JavaScript/199.二叉树的右视图.js
/*
 * @lc app=leetcode.cn id=199 lang=javascript
 *
 * [199] 二叉树的右视图
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
/* 
思路: 第一反应就是层序遍历
*/
/* 
方法一：递归
*/
/* 
Accepted
215/215 cases passed (92 ms)
Your runtime beats 52.79 % of javascript submissions
Your memory usage beats 80.78 % of javascript submissions (39.2 MB)
*/
/* 
var rightSideView = function(root) {
  const res = []
  const bfs = (node, level) => {
    if (!node) return
    res[level] = node.val // 只需要保留最后一个值就OK
    node.left && bfs(node.left, level + 1)
    node.right && bfs(node.right, level + 1)
  }
  bfs(root, 0)
  return res
};
*/


/* 
方法二： 迭代
*/
/* 
Accepted
215/215 cases passed (104 ms)
Your runtime beats 16.78 % of javascript submissions
Your memory usage beats 43.95 % of javascript submissions (39.3 MB)
*/
const rightSideView = root => {
  if (!root) return []
  const res = []
  let level = 0
  const q = [root]
  while (q.length) {
    let len = q.length
    for (let i = 0; i < len; i++) {
      const node = q.shift()
      node.left && q.push(node.left)
      node.right && q.push(node.right)
      res[level] = node.val
    }
    // 层序完了进入下一层
    level++
  }
  return res
}
// @lc code=end

File: Python3/199.二叉树的右视图.py
#
# @lc app=leetcode.cn id=199 lang=python3
#
# [199] 二叉树的右视图
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
"""
深度优先 前序遍历 递归解法
Accepted
215/215 cases passed (36 ms)
Your runtime beats 42.19 % of python3 submissions
Your memory usage beats 6.76 % of python3 submissions (15.1 MB)
"""

"""
class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        res = []
        stack = []

        def handler(node, level):
            if not node:
                return
            if len(stack) <= level:
                stack.append([])
            stack[level].append(node.val)
            handler(node.left, level + 1)
            handler(node.right, level + 1)

        handler(root, 0)

        return [stack[i][-1] for i in range(len(stack))]
"""

"""
深度优先 迭代解法
Accepted
215/215 cases passed (32 ms)
Your runtime beats 69.26 % of python3 submissions
Your memory usage beats 9.67 % of python3 submissions (15.1 MB)
"""


class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        d = {}
        max_level = -1
        stack = [(root, 0)]
        while stack:
            node, level = stack.pop()
            d.setdefault(level, node.val)
            max_level = max(max_level, level)
            if node.left:
                stack.append((node.left, level + 1))
            if node.right:
                stack.append((node.right, level + 1))
        return [d.get(i) for i in range(max_level + 1)]

# @lc code=end
